python/ray/experimental/sgd/interaces.py

```diff
@@ -0,0 +1,134 @@
+            # we still continue training.
+
+class BackendConfig:
+    pass
+
+@dataclass
+class TorchConfig(BackendConfig):
+    address: str # The connection address to use for the process group.
+    process_group_timeout: int # The timeout for process group creation.
+
+@dataclass
+class TFConfig(BackendConfig):
+    pass
+
+
+def setup_torch_process_group(...):
+    pass
+
+def shutdown_torch(...):
+    pass
+
+class TorchExecutor:
+    def __init__(self, worker_args: WorkerArgs, backend_config: TorchConfig,
+                 initialization_hook):
+        self.actor_group = ActorGroup(num_workers=worker_args.num_workers,
+                                  cpus_per_worker=worker_args.num_cpus_per_worker,
+                                  gpus_per_worker=worker_args.num_gpus_per_worker,
+                                  pg_strategy=worker_args.placement_strategy)
+        self.backend_config = backend_config
+        self.initialization_hook = initialization_hook
+
+    def start(self):
+        self.actor_group.start()
+
+        self.actor_group.execute(self.initialization_hook)
+
+        futures = []
+        for i in range(len(self.actor_group.workers)):
+            futures.append(self.actor_group.execute_single_async(i,
+                                                  setup_torch_process_group,
+                                            kwargs={
+                "address": self.backend_config.address,
+                "process_group_timeout": self.backend_config.process_group_timeout,
+                "world_size": len(self.actor_group.workers),
+                "rank": i
+            }))
+
+        ray.get(futures)
+
+    def run(self, training_fn, config):
+        """Runs the training function"""
+
+        # Run the training function asynchronously on all workers.
+        not_ready = self.actor_group.execute_async(training_fn,
+                                                     args=[config])
+
+        # While the training function has not finished on all the workers,
+        # Get results and yield.
+        while not_ready:
+            try:
+                results = self.actor_group.execute(self.fetch_next)
+            except RayActorError:
+                # An actor failed.
+                # handle fault tolerance and elastic training here.
+
+
+            yield results
+            ready, not_ready = ray.wait(not_ready, timeout=0)
+
+    def shutdown(self):
+        self.actor_group.execute(shutodwn_torch)
+        self.actor_group.shutdown()
+
+class Trainer:
+    def __init__(self, worker_args: WorkerArgs, backend: str, backend_config:
+    BackendConfig, initialization_hook: Callable, callbacks: List[
+        "sgd.Callback"]):
+
+        if backend == "torch":
+            self.executor = TorchExecutor(worker_args, backend_config,
+                                          initialization_hook, callbacks)
+
+    def run(self, training_fn, config):
```

Comment:
Random thought: If `training_fn` is a generic `Callable` I'm wondering if there's anything we can/should do to ensure that the function is compatible with the backend.

Reply:
Hmm what were you thinking here? I'm not sure this is possible without actually looking into the function's code. I think we just have to leave it to the user to make sure their training function is compatible with the backend they choose.

Reply:
Yeah it also wouldn't be _required_ for the training function to follow any certain format. We can table this for now and look back at it if there are any user pain points or confusion that we can fix with some sort of systematic validation!